File: easy/0628-maximum-product-of-three-numbers/solution.py
from typing import List


class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        """
        628. Maximum Product of Three Numbers
        Time: O(n)
        Space: O(1)
        """
        max1 = max2 = max3 = float("-inf")  # three largest
        min1 = min2 = float("inf")          # two smallest
        for x in nums:
            if x >= max1:
                max3, max2, max1 = max2, max1, x
            elif x >= max2:
                max3, max2 = max2, x
            elif x > max3:
                max3 = x
            if x <= min1:
                min2, min1 = min1, x
            elif x < min2:
                min2 = x
        return int(max(max1 * max2 * max3, min1 * min2 * max1))

File: easy/0628-maximum-product-of-three-numbers/solution.java
/**
 * 628. Maximum Product of Three Numbers
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int maximumProduct(int[] nums) {
        int max1 = Integer.MIN_VALUE, max2 = Integer.MIN_VALUE, max3 = Integer.MIN_VALUE;
        int min1 = Integer.MAX_VALUE, min2 = Integer.MAX_VALUE;
        for (int x : nums) {
            if (x >= max1) {
                max3 = max2;
                max2 = max1;
                max1 = x;
            } else if (x >= max2) {
                max3 = max2;
                max2 = x;
            } else if (x > max3) {
                max3 = x;
            }
            if (x <= min1) {
                min2 = min1;
                min1 = x;
            } else if (x < min2) {
                min2 = x;
            }
        }
        return Math.max(max1 * max2 * max3, min1 * min2 * max1);
    }
}
